### PR TITLE
Add 'Create New' cpp build configuration 

### PR DESCRIPTION
**Pull request by vince-fugnitto**
Added a `QuickOpen` item to automate the creation of new cpp build configurations

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

File: packages/cpp/src/browser/cpp-build-configurations.ts [hunk 1]
```diff
@@ -54,8 +73,11 @@ export class CppBuildConfigurationManager {
     /** Emitter for when the active build configuration changes.  */
     protected readonly activeConfigChangeEmitter = new Emitter<CppBuildConfiguration | undefined>();
 
+    protected resource: Promise<Resource>;
```
Comment (simark) [inline packages/cpp/src/browser/cpp-build-configurations.ts]: What is this for?  If this is really needed, it would require a better name and a comment.

File: packages/cpp/src/browser/cpp-build-configurations.ts [hunk 2]
```diff
@@ -189,8 +229,9 @@ export class CppBuildConfigurationChanger implements QuickOpenModel {
     }
 
     open() {
+        const configs = this.cppBuildConfigurations.getConfigs();
         this.quickOpenService.open(this, {
-            placeholder: 'Choose a build configuration...',
+            placeholder: (configs.length) ? 'Choose a build configuration...' : 'No build configurations present',
```
Comment (simark) [inline packages/cpp/src/browser/cpp-build-configurations.ts]: Really small nit.  If you only have one entry in settings.json, that still contains the placeholder values, it won't be displayed in the menu.  However, the message shown here will be `'Choose a build configuration...'` rather than `'No build configurations present'`.  The message `'No build configurations present'` should probably be displayed when there are no build configs displayed

File: packages/cpp/src/browser/cpp-build-configurations.ts [hunk 3]
```diff
@@ -33,18 +35,28 @@ class SavedActiveBuildConfiguration {
...
     protected readonly cppPreferences: CppPreferences;
 
+    @inject(EditorManager)
+    protected readonly editorManager: EditorManager;
+
+    @inject(StorageService)
+    protected readonly storageService: StorageService;
+
+    @inject(WorkspacePreferenceProvider)
+    protected readonly preferenceProvider: WorkspacePreferenceProvider;
```
Comment (akosyakov) [inline packages/cpp/src/browser/cpp-build-configurations.ts]: clients should use `PreferenceService`, `PreferenceProvider` is the implementation details

File: packages/cpp/src/browser/cpp-build-configurations.ts [hunk 4]
```diff
@@ -119,6 +131,25 @@ export class CppBuildConfigurationManager {
...
+    }
+
+    /** Create a new build configuration with placeholder values.  */
+    async createConfig(): Promise<void> {
+        const uri = await this.preferenceProvider.getUri();
+        if (uri) {
+            const configs = this.getConfigs();
+            const newConfigs = configs.slice(0);
+            newConfigs.push({ name: namePlaceholder, directory: directoryPlaceholder });
+            await this.preferenceProvider.setPreference(this.BUILD_CONFIGURATIONS_PREFERENCE_KEY, newConfigs);
```
Comment (akosyakov) [inline packages/cpp/src/browser/cpp-build-configurations.ts]: if you use statically-typed proxies for preferences, you don't need constants, ts compiler won't let you to use anything else and will give your proposals

File: packages/cpp/src/browser/cpp-build-configurations.ts (hunk 3 above)
Comment (vince-fugnitto) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: @akosyakov in my implementation I require the use of the `workspace preferences uri` ({workspace}/.theia/settings.json) which was previously provided by the `WorkspacePreferenceProvider`. Do you know of any easy way to obtain the correct uri, or should I instead update both the `PreferenceProvider` and `PreferenceService` to provide a source variable as to where the preferences are located?

**Comment by vince-fugnitto** [comment]
@akosyakov I added a commit that addresses your previous comment of not using the `WorkspacePreferenceProvider` and instead using the service, while making changes to the api to support providing the source. Let me know your thoughts on this approach :)

File: packages/cpp/src/browser/cpp-build-configurations.ts [hunk 5]
```diff
@@ -189,12 +115,28 @@ export class CppBuildConfigurationChanger implements QuickOpenModel {
...
-            placeholder: 'Choose a build configuration...',
+            placeholder: (configs.length) ? 'Choose a build configuration...' : 'No build configurations present',
             fuzzyMatchLabel: true,
             fuzzyMatchDescription: true,
         });
     }
+
+    /** Create a new build configuration with placeholder values.  */
+    async createConfig(): Promise<void> {
+        const resource = await this.getPreferenceProvider(PreferenceScope.Workspace).source;
```
Comment (akosyakov) [inline packages/cpp/src/browser/cpp-build-configurations.ts]: there is `getUri` method, why to expose internals?

Comment (vince-fugnitto) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: @akosyakov I'm not certain as to which `getUri` I should be using to obtain the uri of the workspace preferences. The only method I could find was from the `WorkspacePreferenceProvider` from my initial commit and it was similar to what the `@theia/preferences` uses as well.

https://github.com/theia-ide/theia/blob/f3fb3095ce71e2a0f570d3980f8bce4d704ed8cd/packages/preferences/src/browser/preferences-contribution.ts#L73

Comment (akosyakov) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: `PreferenceProvider` are not supposed to depend on `Resource`, there can be other implementations, so its should be reverted for sure. 

How about we introduce a command to open workspace preferences which will be implemented by `@theia/preferences` extension and you call it from here?

Also please if you change some APIs assign people who designed this APIs for reviews.

Comment (akosyakov) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: For example `PreferenceProvider` based on the schema: https://github.com/theia-ide/theia/pull/2701/files#diff-efc63fa928532aa0ab10afc95417b026R53

Comment (vince-fugnitto) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: Thank you and yes I will be reverting the middle commit for sure and use a different approach based on your feedback. Sorry about not adding the people who designed the API for review, changes were initially self-contained within `@theia/cpp` and quickly grew.

**Comment by simark** [comment]
Parallel to Anton's comments: I think it would be more natural to have `interface CppBuildConfiguration` in the `cpp-build-configurations.ts` file, along with things that are more core to the concept.  The `CppBuildConfigurationManager` class can be in there, or in its own `cpp-build-configurations-manager.ts` file.  Everything UI-related could go in `cpp-build-configurations-ui.ts`.

File: packages/cpp/src/browser/cpp-build-configurations.ts (hunk 5 above)
Comment (marechal-p) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: @akosyakov the issue here is that in order to open the edited preference file, we need to get its URI, and if we solely use a `PreferenceProvider`, there is no way to guess where it got its preferences from:

https://github.com/theia-ide/theia/blob/master/packages/core/src/browser/preferences/preference-provider.ts#L46-L52

We have get/set but no way to query the provider in order to get something representing its source of preferences.

Same goes with the recently added preference widget:

https://github.com/theia-ide/theia/blob/master/packages/preferences/src/browser/preferences-contribution.ts#L40-L41

In the preferences we also kindof refer directly to the implementation detail for the user and workspace preferences. I'm not fond of it either, but it works so I'm ok.

---

In short: How should we fetch the URI of the file to open in the editor when editing the preferences?

Comment (akosyakov) [reply packages/cpp/src/browser/cpp-build-configurations.ts]: I suggest that we move `Open Preferences` command declaration to core extension and keep its handler in preferences extension.

From here, you do something like:
```ts
commands.executeCommand(PREFERENCES_COMMAND.OPEN.id, {
    scope: PreferenceScope.WORKSPACE 
});
````

**Comment by vince-fugnitto** [comment]
@akosyakov are you okay with the following changes to the PR?

**Comment by lmcbout** [comment]
It looks like if the file containing the new definition of the build config (settings.json) is **dirty**, nothing is being created in the file, no warning to the end user and an error is created on the server console.log:

root ERROR C/C++: I[13:14:58.326] <-- $/setTraceNotification
E[13:14:58.326] Error -32601: method not found

**Comment by akosyakov** [comment]
@vince-fugnitto looking now

File: packages/preferences/src/browser/preferences-contribution.ts [hunk 6]
```diff
@@ -81,4 +92,18 @@ export class PreferencesContribution extends AbstractViewContribution<Preference
         super.openView({ activate: true });
     }
 
+    /** Open User Preferences with Editor */
+    protected async openUserPreferences(): Promise<void> {
+        const uri = await this.userPreferenceProvider.getUri();
+        this.editorManager.open(uri);
```
[the remark this answers is not in the document]
Comment (vince-fugnitto) [reply packages/preferences/src/browser/preferences-contribution.ts]: Yes I think that would be a more appropriate solution and I won't need to re-implement error handling when the preferences are `dirty` and the user attempts to add a new configuration. I'll make the changes and test.

[the remark this answers is not in the document]
Comment (vince-fugnitto) [reply packages/preferences/src/browser/preferences-contribution.ts]: do you know of an easy way to be able to have the `workspace preferences` tab selected when opening the preference widget ?

[the remark this answers is not in the document]
Comment (akosyakov) [reply packages/preferences/src/browser/preferences-contribution.ts]: We will need to add a new API to the preference widget to allow it.

**Comment by vince-fugnitto** [comment]
@akosyakov I have updated the PR to use the `Preferences Widget` and set the active editor based on scope

File: packages/preferences/src/browser/preferences-contribution.ts [hunk 7]
```diff
@@ -49,21 +47,21 @@ export class PreferencesContribution extends AbstractViewContribution<Preference
...
         menus.registerMenuAction(CommonMenus.FILE_SETTINGS_SUBMENU_OPEN, {
-            commandId: PREFERENCES_COMMAND.id,
+            commandId: CommonCommands.OPEN_PREFERENCES.id,
             order: 'a10'
         });
     }
 
-    protected async openPreferences(): Promise<void> {
+    protected async openPreferences(preferenceScope: PreferenceScope): Promise<void> {
         const userUri = this.userPreferenceProvider.getUri();
```
[the remark this answers is not in the document]
Comment (vince-fugnitto) [reply packages/preferences/src/browser/preferences-contribution.ts]: I made the changes, and confirmed that it isn't necessary to handle creating the `user preferences`, but the creation of `workspace preferences` is in fact needed.

File: packages/preferences/src/browser/preferences-contribution.ts [hunk 8]
```diff
@@ -49,33 +45,30 @@ export class PreferencesContribution extends AbstractViewContribution<Preference
...
+    protected async openPreferences(preferenceScope: PreferenceScope): Promise<void> {
         const wsUri = await this.workspacePreferenceProvider.getUri();
         if (wsUri && !await this.filesystem.exists(wsUri.toString())) {
             await this.filesystem.createFile(wsUri.toString());
         }
 
+        const widget = await this.widget;
+        widget.preferenceScope = preferenceScope;
         super.openView({ activate: true });
+        widget.activatePreferenceEditor(preferenceScope);
```
Comment (akosyakov) [inline packages/preferences/src/browser/preferences-contribution.ts]: I think it should be an internal API, and be called `onActivate` callback by the widget itself, i.e. from the client perspective calling `widget.preferenceScope = preferenceScope` should be enough, the client should not care to calls something else. I'm fine to change it later though.

File: packages/preferences/src/browser/preferences-tree-widget.ts [hunk 9]
```diff
@@ -171,6 +185,12 @@ export class PreferencesContainer extends SplitPanel implements ApplicationShell
         super.onCloseRequest(msg);
         this.dispose();
     }
+
+    public activatePreferenceEditor(preferenceScope: PreferenceScope) {
```
Comment (akosyakov) [inline packages/preferences/src/browser/preferences-tree-widget.ts]: `public` is the default, I think it should be internal and called from `onActivateRequest` to be executed on each activation

File: packages/preferences/src/browser/preferences-contribution.ts (hunk 8 above)
Comment (vince-fugnitto) [reply packages/preferences/src/browser/preferences-contribution.ts]: the only reason I was calling it also at this point was so that if the preferences widget was already open (`user preferences`), and the user attempted to add a new cpp config, the editor would toggle to the `workspace preferences` where they can see the new configuration.

**Comment by vince-fugnitto** [comment]
@akosyakov thanks for all the reviews and feedback you provided throughout the PR!
